crates/iceberg/src/spec/values.rs

```diff
@@ -0,0 +1,417 @@
+    /// 16-byte big-endian value
+    UUID(i128),
+    /// Binary value
+    Fixed(usize, Vec<u8>),
+    /// Binary value (without length)
+    Binary(Vec<u8>),
+    /// Stores unscaled value as two’s-complement big-endian binary,
+    /// using the minimum number of bytes for the value
+    Decimal(Decimal),
+    /// A struct is a tuple of typed values. Each field in the tuple is named and has an integer id that is unique in the table schema.
+    /// Each field can be either optional or required, meaning that values can (or cannot) be null. Fields may be any type.
+    /// Fields may have an optional comment or doc string. Fields can have default values.
+    Struct(Struct),
+    /// A list is a collection of values with some element type.
+    /// The element field has an integer id that is unique in the table schema.
+    /// Elements can be either optional or required. Element types may be any type.
+    List(Vec<Option<Value>>),
+    /// A map is a collection of key-value pairs with a key type and a value type.
+    /// Both the key field and value field each have an integer id that is unique in the table schema.
+    /// Map keys are required and map values can be either optional or required. Both map keys and map values may be any type, including nested types.
+    Map(HashMap<String, Option<Value>>),
+}
+
+impl From<Value> for ByteBuf {
+    fn from(value: Value) -> Self {
+        match value {
+            Value::Boolean(val) => {
+                if val {
+                    ByteBuf::from([0u8])
+                } else {
+                    ByteBuf::from([1u8])
+                }
+            }
+            Value::Int(val) => ByteBuf::from(val.to_le_bytes()),
+            Value::LongInt(val) => ByteBuf::from(val.to_le_bytes()),
+            Value::Float(val) => ByteBuf::from(val.to_le_bytes()),
+            Value::Double(val) => ByteBuf::from(val.to_le_bytes()),
+            Value::Date(val) => ByteBuf::from(val.to_le_bytes()),
+            Value::Time(val) => ByteBuf::from(val.to_le_bytes()),
+            Value::Timestamp(val) => ByteBuf::from(val.to_le_bytes()),
+            Value::TimestampTZ(val) => ByteBuf::from(val.to_le_bytes()),
+            Value::String(val) => ByteBuf::from(val.as_bytes()),
+            Value::UUID(val) => ByteBuf::from(val.to_be_bytes()),
+            Value::Fixed(_, val) => ByteBuf::from(val),
+            Value::Binary(val) => ByteBuf::from(val),
+            _ => todo!(),
+        }
+    }
+}
+
+/// The partition struct stores the tuple of partition values for each file.
+/// Its type is derived from the partition fields of the partition spec used to write the manifest file.
+/// In v2, the partition struct’s field ids must match the ids from the partition spec.
+#[derive(Debug, Clone, PartialEq)]
+pub struct Struct {
+    /// Vector to store the field values
+    fields: Vec<Option<Value>>,
+    /// A lookup that matches the field name to the entry in the vector
+    lookup: HashMap<String, usize>,
```

Comment:
I would move the lookups to the schema

Reply:
This is currently because I tried to serialize/deserialize the `partition` struct like in the `struct_value` test. I'm not sure if this is the right approach. Since the partition struct is stored as a avro record which maps names to values I needed something like a `Hashmap<String,Value>`. However serializing/deserializing a hashmap somehow didn't work with avro since the record is stored as a `Vec<(Key,Value)>`, where the tuples can be in different order everytime. And the order seemed to play a role. So I had the feeling that the tuple order could be different everytime, which means you cannot use a global schema lookup to get the fields.

That being said, I think using serde to serialize/deserialize the `partition` struct might not be the best idea.

I'm struggling anyway with the different serialization schemes for binary and json data. @Fokko maybe you could share your experience with python.

Reply:
I see, thanks for the context.

So in Python we deserialize the `102: partition struct<...>` in a `Record`, that implement `StructProtocol`. Since we have the schema from the Avro file, we know what values are at which positions, and we can lookup at the index (which is super fast! :)

So let's say that you have a `day(dt) as partition_by_day` partitioning. Where `dt` is a `Timestamp` in your schema, and `partition_by_day` is the hidden partition that buckets the writes per file. Then the partition struct would look like `struct<partition_by_day: int>`. If the user does an expression `dt >= '1990-03-01 19:25:00'`, which is read as a string value, but gets converted to a date value when binding the expression to the schema.

Similar to binding a normal expression to a Schema, we also bind the hidden partition filter to the `partition` schema. Since we know the position of the field in the schema, we can just evaluate using an accessor at index 0 in this case. We take the value and using the evaluator, we check if it might match, if that's the case, we include the file in the scan.

Code-wise, this is all done in the [ManifestEvaluator](https://github.com/apache/iceberg/blob/master/python/pyiceberg/expressions/visitors.py#L783-L789)